### ISSUE TITLE
A very strange bug with tf.cond, update_ops and global_step

### ISSUE THREAD

**Issue by x10000year** [tensorflow/tensorflow#14699]
I'm using tf.Estimator with custom model_fn. When training, the estimator usually outputs log like:

INFO:tensorflow:loss = 1109.14, step = 1
INFO:tensorflow:loss = 937.876, step = 101 (6.245 sec)
INFO:tensorflow:loss = 632.192, step = 201 (6.195 sec)

By default, the printed steps should be 1, 101, 201... However, when I use the following function (which is simplified to reproduce the bug) in any place of the model:

```
def my_op(inputs, name=None):
  with tf.variable_scope(name, default_name='my_scope', reuse=False):
    count = tf.get_variable('count', shape=[],
      initializer=tf.zeros_initializer(), trainable=False)

    def myfunc1():
      return 1

    def myfunc2():
      tf.add_to_collection(tf.GraphKeys.UPDATE_OPS, count.assign_add(10.0))
      return 1

    tf.cond(tf.less(1, 2), myfunc1, myfunc2)
    return inputs

```

The log becomes something like:

INFO:tensorflow:loss = 1130.58, step = 0
INFO:tensorflow:loss = 940.298, step = 0 (6.352 sec)

The global step is always 0. After some tests, I found that the global step is not updated if myfunc2 is not executed. For example, if I write
`    tf.cond(tf.less(count, 2), myfunc1, myfunc2)`
then the global step is always 1.

I suspect that this is caused by the tf.add_to_collection(tf.GraphKeys.UPDATE_OPS, ...) in myfunc2, in which my intend is to update a variable when some condition holds. Maybe op created inside tf.cond can not be used as a dependency outside, but no error message is reported.

If I cannot add ops to UPDATE_OPS inside tf.cond, does this imply that stateful operations (like tf.layers.batch_normalization) can not be used inside tf.cond? So I cannot dynamically choose a network module from a set of network modules to execute if the network modules use any stateful operations like tf.layers.batch_normalization?

my tensorflow version: ('v1.4.0-rc0-10-g756a7fc', '1.4.0-rc1')

**Comment by JeremyCCHsu** [comment]
I have the same problem while using `tf.train.Supervisor` in TF nightly build 1118. 
1. `global_step` was explicitly passed to the supervisor, but is not updated (so every summary collapsed to one point).
2. Models were saved in the same name.

My code was exactly the same as the one I was using in TF 1.4.0-rc0.

**Comment by tatatodd** [comment]
@skye might have some ideas behind the interaction between `tf.cond` and the global step.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by Multihuntr** [comment]
I was unable to reproduce this bug given the information provided. The following script runs as expected. 

``` {py}
import numpy as np
import tensorflow as tf

def my_op(inputs, name=None):
  with tf.variable_scope(name, default_name='my_scope', reuse=False):
    count = tf.get_variable('count', shape=[],
      initializer=tf.zeros_initializer(), trainable=False)

    def myfunc1():
      return 1

    def myfunc2():
      tf.add_to_collection(tf.GraphKeys.UPDATE_OPS, count.assign_add(10.0))
      return 1

    tf.cond(tf.less(count, 2), myfunc1, myfunc2)
    return inputs

global_step = tf.get_variable('global_step', [],
  initializer=tf.zeros_initializer(), trainable=False, dtype=tf.int32)
a = tf.placeholder(tf.float32, [None, 5])
b = tf.placeholder(tf.float32, [None, 3])
a_out = tf.layers.dense(my_op(a, 'somewhere'), 3)

loss = tf.losses.mean_squared_error(b, a_out)
opt = tf.train.AdamOptimizer(learning_rate=1e-2)

train_op = opt.minimize(loss, global_step)

with tf.Session() as sess:
  sess.run(tf.global_variables_initializer())

  print(sess.run(global_step)) # prints 0
  sess.run(train_op, { a: [[0.1, 0.2, 0.3, 0.4, 0.5]], b: [[1., 2., 3.]] })
  print(sess.run(global_step)) # prints 1
  sess.run(train_op, { a: [[0.1, 0.2, 0.3, 0.4, 0.5]], b: [[1., 2., 3.]] })
  print(sess.run(global_step)) # prints 2
  sess.run(train_op, { a: np.empty(shape=[0,5]), b: np.empty(shape=[0,3]) })
  print(sess.run(global_step)) # prints 3
```

As does changing the `tf.less(count, 2)` to `tf.less(1,2)`.
As does using `a` directly in the `tf.layers.dense`, but still calling `my_op(a, 'somewhere')` elsewhere.

So there's another piece to this puzzle; there's something else going on to cause the issue.

**Comment by Multihuntr** [comment]
I figured out the missing ingredient to reproduce the bug: you need to put your `opt.minimize` inside a `with tf.control_dependencies(tf.get_collection(tf.GraphKeys.UPDATE_OPS))`

Turns out this is a rather dangerous bug: Adding an operation to a collection inside a function used in a `tf.cond` **silently** prevents any operations that use that collection as a control dependency, regardless of whether that `tf.cond` is ever used.

So, it's not just the global_step; it's everything! Here's the minimal code I put together to highlight the bug in action.

``` 
import numpy as np
import tensorflow as tf

flag = tf.placeholder(tf.bool, [])
global_step = tf.train.get_or_create_global_step()

# Include breaking code, putting a no_op inside a cond,
# which isn't even used anywhere, and shouldn't impact anything else
def myfunc():
  tf.add_to_collection(tf.GraphKeys.UPDATE_OPS, tf.no_op())
  return 1
tf.cond(flag, myfunc, lambda: 1)

# Make our ad hoc model
a = tf.placeholder(tf.float32, [None, 5])
b = tf.placeholder(tf.float32, [None, 3])
a_out = tf.layers.dense(a, 3)
graph = tf.get_default_graph()
bias = graph.get_tensor_by_name('dense/bias:0')
weights = graph.get_tensor_by_name('dense/kernel:0')

# Define our train_op
loss = tf.losses.mean_squared_error(b, a_out)
opt = tf.train.GradientDescentOptimizer(learning_rate=1e-2)
with tf.control_dependencies(tf.get_collection(tf.GraphKeys.UPDATE_OPS)):
  train_op = opt.minimize(loss, global_step)

# Test everything
with tf.Session() as sess:
  sess.run(tf.global_variables_initializer())

  print('BEFORE ANYTHING IS RUN')
  print(sess.run(global_step))
  print(sess.run(bias))
  print(sess.run(weights))
  sess.run(train_op, { a: [[0.1, 0.2, 0.3, 0.4, 0.5]], b: [[1., 2., 3.]], flag: False })
  print('AFTER FIRST RUN:  we do not follow the path with the update op added')
  print(sess.run(global_step))
  print(sess.run(bias))
  print(sess.run(weights))
  sess.run(train_op, { a: [[0.1, 0.2, 0.3, 0.4, 0.5]], b: [[1., 2., 3.]], flag: True })
  print('AFTER SECOND RUN: we follow the path with the update op added')
  print(sess.run(global_step))
  print(sess.run(bias))
  print(sess.run(weights))
```

Running the above code gives

```
BEFORE ANYTHING IS RUN
0
[ 0.  0.  0.]
[[-0.37039062 -0.37549886  0.73892838]
 [-0.20615649  0.03782606  0.50503689]
 [-0.43216512  0.73242658  0.42330664]
 [ 0.11180687  0.07816333 -0.02583456]
 [ 0.15668219  0.58735186  0.83961934]]
AFTER FIRST RUN:  we do not follow the path with the update op added
0
[ 0.  0.  0.]
[[-0.37039062 -0.37549886  0.73892838]
 [-0.20615649  0.03782606  0.50503689]
 [-0.43216512  0.73242658  0.42330664]
 [ 0.11180687  0.07816333 -0.02583456]
 [ 0.15668219  0.58735186  0.83961934]]
AFTER SECOND RUN: we follow the path with the update op added
1
[ 0.00723237  0.0099021   0.01525755]
[[-0.36966738 -0.37450865  0.74045414]
 [-0.20471002  0.03980648  0.50808841]
 [-0.42999542  0.73539722  0.42788389]
 [ 0.11469982  0.08212417 -0.01973154]
 [ 0.16029838  0.59230292  0.84724814]]
```

I looked around for other people experiencing problems with adding ops to the UPDATE_OPS collection inside a `tf.cond` and couldn't find anything. The fact that it completely halts training despite not being a required node for the `train_op` and without any errors means that you could silently break your whole training with just those few lines.

**Comment by skye** [comment]
@Multihuntr thank you for digging into this! I have a large backlog of issues to diagnose so I can't look into a potential fix for this right now, so for now it's probably best to avoid adding to collections inside a cond.

**Comment by Multihuntr** [comment]
I've played around with it a little more, and it turns out that this issue only occurs when the operation is actually created inside the cond. So this will work fine:

```
a_no_op = tf.no_op()
def myfunc():
  tf.add_to_collection(tf.GraphKeys.UPDATE_OPS, a_no_op)
  return 1
tf.cond(tf.less(2,1), myfunc, lambda: 1)
```

But this will break everything:

```
def myfunc():
  tf.add_to_collection(tf.GraphKeys.UPDATE_OPS, tf.no_op())
  return 1
tf.cond(tf.less(2,1), myfunc, lambda: 1)
```

This is because the operations defined outside the `myfunc` are always executed, regardless of the result of the condition. So, you CAN use batch norm with conditionals, you just have to define the branch containing the batch norm outside the `tf.cond`. However this means that that branch will always execute. This is definitely a workaround that will only work in some cases (e.g. it does not solve the OP's problem, as the assignment operation should only be executed once).

I'll definitely avoid this, but the biggest problem I can see is that it's silent and extremely easy to accidentally cause (using batch norm in a conditional is perfectly reasonable). I wonder how many people have been stung by this and not even known that it was a bug within tensorflow itself.

Is there some way we can make this bug more visible? Hotfix the docs for tf.cond or something?

**Comment by Multihuntr** [comment]
A better solution for using batch norm inside a conditional: [`tf.contrib.layers.batch_normalization`](https://www.tensorflow.org/api_docs/python/tf/contrib/layers/batch_norm) allows you to set `updates_collections=None` to force the updates to happen before you get the batch norms output (and thus not adding them to UPDATE_OPS).

However [`tf.layers.batch_normalization`](https://www.tensorflow.org/api_docs/python/tf/layers/batch_normalization) doesn't seem to have this option.

If you use Sonnet, then when constructing the batch norm module, you can use `update_ops_collection=None` ([source](https://github.com/deepmind/sonnet/blob/master/sonnet/python/modules/batch_norm.py#L172)).

**Comment by Multihuntr** [comment]
Ah, that's actually a fairly general workaround. Instead of adding ops to the UPDATE_OPS, you enforce that it must be completed before getting the output of your branch in the `tf.cond`.

So the solution to the OPs problem is to rewrite their `myfunc2` as follows:

```
  def myfunc2():
      with tf.control_dependances([count.assign_add(10.0)]):
        return 1
```

**Comment by tensorflowbutler** [comment]
A member of the TensorFlow organization has replied after the stat:awaiting tensorflower label was applied.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @skye: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @skye: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @skye: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @skye: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by ppwwyyxx** [comment]
The similar issue has appeared several times in #18795, #14809. I think it's a good feature to add an extra option to `layers.batch_normalization`, similar to that of `contrib.batch_norm`.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @skye: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @skye: It has been 30 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @skye: It has been 45 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @skye: It has been 60 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @skye: It has been 75 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @skye: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @skye: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @skye: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @skye: It has been 59 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @skye: It has been 74 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @skye: It has been 89 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @skye: It has been 104 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @skye: It has been 119 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @skye: It has been 134 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by jvishnuvardhan** [comment]
This is a stale issue. Please check the issue with latest TensorFlow. If the issue still persists in the newer version of TF, please feel free to reopen it by providing details about the issue and a standalone code to reproduce the issue. Thanks!

**Comment by Multihuntr** [comment]
I ran my test code again using the `tensorflow/tensorflow:1.13.2-gpu-py3` docker image and got this result:

```
BEFORE ANYTHING IS RUN
0
[0. 0. 0.]
[[ 0.3860709   0.45507103 -0.1816141 ]
 [-0.00840878  0.7818753  -0.605694  ]
 [-0.04105538 -0.03619885 -0.10100174]
 [ 0.8570104   0.44614607  0.7425906 ]
 [ 0.7079007   0.8293366  -0.6455219 ]]
AFTER FIRST RUN:  we do not follow the path with the update op added
0
[0. 0. 0.]
[[ 0.3860709   0.45507103 -0.1816141 ]
 [-0.00840878  0.7818753  -0.605694  ]
 [-0.04105538 -0.03619885 -0.10100174]
 [ 0.8570104   0.44614607  0.7425906 ]
 [ 0.7079007   0.8293366  -0.6455219 ]]
AFTER SECOND RUN: we follow the path with the update op added
1
[0.00185758 0.00810567 0.02130217]
[[ 0.38625666  0.4558816  -0.17948389]
 [-0.00803727  0.78349644 -0.6014336 ]
 [-0.04049811 -0.03376715 -0.09461109]
 [ 0.85775346  0.44938833  0.75111145]
 [ 0.70882946  0.8333894  -0.63487077]]
```

This issue still exists in the latest version of Tensorflow. When the `sess.run` does not execute the `tf.no_op`, the `opt.minimize` operation silently fails to run.

I do not have authority to re-open this issue.
